Common/Thread/ParallelLoop.cpp

```diff
@@ -0,0 +1,142 @@
+#include <cstring>
+
+#include "Common/Thread/ParallelLoop.h"
+#include "Common/CPUDetect.h"
+
+class LoopRangeTask : public Task {
+public:
+	LoopRangeTask(WaitableCounter *counter, const std::function<void(int, int)> &loop, int lower, int upper)
+		: counter_(counter), loop_(loop), lower_(lower), upper_(upper) {}
+
+	void Run() override {
+		loop_(lower_, upper_);
+		counter_->Count();
+	}
+
+	std::function<void(int, int)> loop_;
+	WaitableCounter *counter_;
+
+	int lower_;
+	int upper_;
+};
+
+WaitableCounter *ParallelRangeLoopWaitable(ThreadManager *threadMan, const std::function<void(int, int)> &loop, int lower, int upper, int minSize) {
+	if (minSize == -1) {
+		minSize = 1;
+	}
+
+	int numTasks = threadMan->GetNumLooperThreads();
+
+	int range = upper - lower;
+	if (range <= 0) {
+		// Bad range. A finished counter allocated.
+		return new WaitableCounter(0);
+	}
+
+	if (range <= numTasks) {
+		// Just assign one task per thread, as many as we have.
+		WaitableCounter *waitableCounter = new WaitableCounter(range);
+		for (int i = 0; i < range; i++) {
+			threadMan->EnqueueTaskOnThread(i, new LoopRangeTask(waitableCounter, loop, i, i + 1), TaskType::CPU_COMPUTE);
+		}
+		return waitableCounter;
+	} else {
+		// Split the range between threads. Allow for some fractional bits.
+		const int fractionalBits = 8;
+
+		int64_t totalFrac = (int64_t)range << fractionalBits;
+		int64_t delta = totalFrac / numTasks;
+
+		delta = std::max(delta, (int64_t)minSize << fractionalBits);
+
+		// Now we can compute the actual number of tasks.
+		// Remember that stragglers are done on the current thread
+		// so we don't round up.
+		numTasks = (int)(totalFrac / delta);
+		printf("numTasks: %d\n", numTasks);
+
+		WaitableCounter *waitableCounter = new WaitableCounter(numTasks);
+		int64_t counter = (int64_t)lower << fractionalBits;
+
+		// Split up tasks as equitable as possible.
+		for (int i = 0; i < numTasks; i++) {
```

Comment:
Due to the fractional bits, we'll end up doing most of the tasks on the threads, the downward rounding that caused previously made the remainder work more even.  I suppose we're using all the threads here, so we'll get a spike of i.e. 4 + 1 on a quad core, but it'll be a small remainder and most will wait for a thread switch...

My slight ideal is that the remainder (calling) thread has enough that ~75% of the time when it checks if the other tasks are complete, they just completed beforehand.  Thus `- 1` here.  But maybe I'm trying too hard to avoid the thread switch, it was more expensive before...

-[Unknown]

Reply:
Actually, in the case where the minSize is large, we can still end up with a large remainder, think 0-100 with a minSize of 40. 20 items end up in the remainder.

Though I guess that's not the most common case...

But I think the thread switch overhead etc is really not a big deal anyway anymore, no. 

We can play around with this later and see if we can find a case where it makes a difference.